File: web/src/server/anthropic.ts
import "server-only";

import Anthropic from "@anthropic-ai/sdk";
import type { AIProvider, ChatMessage, ToolDefinition, ToolStreamChunk } from "./ai-provider";

export const DEFAULT_SYSTEM_PROMPT = `You are a transit route planning assistant for Toronto.

You help urban planners design new transit lines. When the user describes a route requirement,
respond conversationally and helpfully. If they ask you to generate a specific route, also output
a JSON block at the end of your message in this exact format:

\`\`\`route
{
  "name": "Route Name",
  "type": "subway" | "streetcar" | "bus",
  "color": "#hexcolor",
  "stops": [
    { "name": "Stop Name", "coords": [-79.3832, 43.6532] }
  ]
}
\`\`\`

Coordinates are [longitude, latitude] in WGS84. Only include the JSON block when generating
an actual route. Use realistic Toronto coordinates. Keep stop names concise.`;

type StoredAssistant = {
  name: string;
  systemPrompt: string;
};

type StoredMessage = {
  role: "user" | "assistant";
  content: string;
};

type StoredThread = {
  assistantId: string;
  messages: StoredMessage[];
};

// 📖 Learn: a "factory function" creates and returns an object. The Maps
// (assistantStore, threadStore) are closed over — each call to
// createAnthropicProvider() gets its own private store.
export function createAnthropicProvider(): AIProvider {
  if (!process.env.ANTHROPIC_API_KEY) {
    throw new Error("ANTHROPIC_API_KEY environment variable is required");
  }

  const client = new Anthropic({ apiKey: process.env.ANTHROPIC_API_KEY });
  const assistantStore = new Map<string, StoredAssistant>();
  const threadStore = new Map<string, StoredThread>();

  function getAssistant(id: string): StoredAssistant {
    const a = assistantStore.get(id);
    if (!a) throw new Error(`Unknown assistant: ${id}`);
    return a;
  }

  function getThread(id: string): StoredThread {
    const t = threadStore.get(id);
    if (!t) throw new Error(`Unknown thread: ${id}`);
    return t;
  }

  function extractText(content: Anthropic.Messages.Message["content"]): string {
    return content
      .filter((block): block is Anthropic.TextBlock => block.type === "text")
      .map((block) => block.text)
      .join("");
  }

  return {
    async createAssistant(name, systemPrompt = DEFAULT_SYSTEM_PROMPT) {
      const assistantId = crypto.randomUUID();
      assistantStore.set(assistantId, { name, systemPrompt });
      return assistantId;
    },

    async createThread(assistantId) {
      getAssistant(assistantId);
      const threadId = crypto.randomUUID();
      threadStore.set(threadId, { assistantId, messages: [] });
      return threadId;
    },

    async *streamMessage(threadId, content, model = "claude-haiku-4-5-20251001", maxTokens = 600) {
      const thread = getThread(threadId);
      const assistant = getAssistant(thread.assistantId);
      const nextMessages: StoredMessage[] = [
        ...thread.messages,
        { role: "user", content },
      ];

      const stream = client.messages.stream({
        model,
        max_tokens: maxTokens,
        system: assistant.systemPrompt,
        messages: nextMessages,
      });

      let full = "";
      try {
        for await (const event of stream) {
          if (
            event.type === "content_block_delta" &&
            event.delta.type === "text_delta"
          ) {
            full += event.delta.text;
            yield event.delta.text;
          }
        }
      } finally {
        if (full) {
          threadStore.set(threadId, {
            ...thread,
            messages: [...nextMessages, { role: "assistant", content: full }],
          });
        }
      }
    },

    // 📖 Learn: "tool_choice: { type: 'tool', name: '...' }" tells Claude it MUST call
    // that specific tool. The model can still write text first (its reasoning), then
    // calls the tool. We stream both: text chunks come as text_delta events; the tool
    // arguments arrive as input_json_delta fragments that we reassemble into JSON.
    async *streamMessageWithTool(threadId, content, tool: ToolDefinition, model = "claude-haiku-4-5-20251001", maxTokens = 900): AsyncGenerator<ToolStreamChunk> {
      const thread = getThread(threadId);
      const assistant = getAssistant(thread.assistantId);
      const nextMessages: StoredMessage[] = [
        ...thread.messages,
        { role: "user", content },
      ];

      const stream = client.messages.stream({
        model,
        max_tokens: maxTokens,
        system: assistant.systemPrompt,
        messages: nextMessages,
        tools: [{
          name: tool.name,
          description: tool.description,
          // 📖 Learn: Anthropic names this field "input_schema" (snake_case)
          // even though we store it as "inputSchema" in our shared ToolDefinition.
          input_schema: tool.inputSchema as Anthropic.Messages.Tool["input_schema"],
        }],
        // Force Claude to call exactly this tool (not just "maybe use a tool")
        tool_choice: { type: "tool", name: tool.name },
      });

      let fullText = "";
      let toolInputJson = "";

      try {
        for await (const event of stream) {
          if (event.type === "content_block_delta") {
            if (event.delta.type === "text_delta") {
              fullText += event.delta.text;
              yield { type: "text", text: event.delta.text };
            } else if (event.delta.type === "input_json_delta") {
              // 📖 Learn: the tool JSON is streamed as fragments (e.g. '{"name"' then
              // ':"Eglinton"' then ',"stops":[' ...). We accumulate and parse at the end.
              toolInputJson += event.delta.partial_json;
            }
          }
        }
      } finally {
        // Store only the text portion — agents run once per council, so we don't need
        // the tool call in history for multi-turn continuity.
        if (fullText) {
          threadStore.set(threadId, {
            ...thread,
            messages: [...nextMessages, { role: "assistant", content: fullText }],
          });
        }
      }

      if (toolInputJson) {
        try {
          yield { type: "tool", input: JSON.parse(toolInputJson) as Record<string, unknown> };
        } catch {
          // Malformed JSON from the model — caller will treat route as null
        }
      }
    },

    async sendMessage(threadId, content, model = "claude-haiku-4-5-20251001", maxTokens = 600) {
      const thread = getThread(threadId);
      const assistant = getAssistant(thread.assistantId);
      const nextMessages: StoredMessage[] = [
        ...thread.messages,
        { role: "user", content },
      ];

      const response = await client.messages.create({
        model,
        max_tokens: maxTokens,
        system: assistant.systemPrompt,
        messages: nextMessages,
      });

      const text = extractText(response.content);
      threadStore.set(threadId, {
        ...thread,
        messages: [...nextMessages, { role: "assistant", content: text }],
      });
      return text;
    },

    async *streamDirect(system, messages: ChatMessage[], model = "claude-haiku-4-5-20251001", maxTokens = 1024) {
      // 📖 Learn: we pass the full message list directly (no assistant/thread lookup)
      // so the caller controls history. Anthropic accepts this format natively.
      const stream = client.messages.stream({
        model,
        max_tokens: maxTokens,
        system,
        messages,
      });

      for await (const event of stream) {
        if (
          event.type === "content_block_delta" &&
          event.delta.type === "text_delta"
        ) {
          yield event.delta.text;
        }
      }
    },
  };
}

File: python_server/api/anthropic.py
from __future__ import annotations

import json
import os
import uuid
from dataclasses import dataclass
from pathlib import Path
from typing import Any, AsyncIterator, Awaitable, Callable

import anthropic
from dotenv import load_dotenv

load_dotenv(Path(__file__).resolve().parents[2] / ".env", override=False)

CLAUDE_KEY = (
    os.environ.get("ANTHROPIC_API_KEY", "")
    or os.environ.get("CLAUDE_KEY", "")
    or os.environ.get("CLAUDE_API_KEY", "")
)

SYSTEM_PROMPT = """You are a transit route planning assistant for Toronto.

You help urban planners design new transit lines. When the user describes a route requirement,
respond conversationally and helpfully. If they ask you to generate a specific route, also output
a JSON block at the end of your message in this exact format:

```route
{
  "name": "Route Name",
  "type": "subway" | "streetcar" | "bus",
  "color": "#hexcolor",
  "stops": [
    { "name": "Stop Name", "coords": [-79.3832, 43.6532] }
  ]
}
```

Coordinates are [longitude, latitude] in WGS84. Only include the JSON block when generating
an actual route. Use realistic Toronto coordinates. Keep stop names concise.
"""


@dataclass
class ToolCallEvent:
    name: str
    input: dict
    call_id: str
    result: Any


_assistants: dict[str, str] = {}
_threads: dict[str, dict[str, Any]] = {}


def _require_key() -> None:
    if not CLAUDE_KEY:
        raise RuntimeError("ANTHROPIC_API_KEY is required for the Python AI server")


async def create_assistant(name: str, system_prompt: str) -> str:
    _require_key()
    assistant_id = str(uuid.uuid4())
    _assistants[assistant_id] = system_prompt
    return assistant_id


async def create_thread(assistant_id: str) -> str:
    _require_key()
    thread_id = str(uuid.uuid4())
    _threads[thread_id] = {
        "assistant_id": assistant_id,
        "system_prompt": _assistants.get(assistant_id, ""),
        "messages": [],
    }
    return thread_id


async def stream_message(
    thread_id: str,
    content: str,
    model: str = "claude-haiku-4-5-20251001",
    max_tokens: int = 600,
) -> AsyncIterator[str]:
    _require_key()

    thread = _threads.setdefault(
        thread_id,
        {"assistant_id": None, "system_prompt": "", "messages": []},
    )
    next_messages = list(thread["messages"]) + [{"role": "user", "content": content}]
    client = anthropic.AsyncAnthropic(api_key=CLAUDE_KEY)

    full_response = ""
    async with client.messages.stream(
        model=model,
        max_tokens=max_tokens,
        system=thread["system_prompt"] or None,
        messages=next_messages,
    ) as stream:
        async for text in stream.text_stream:
            full_response += text
            yield text

    thread["messages"] = next_messages + [{"role": "assistant", "content": full_response}]


async def stream_message_with_tools(
    thread_id: str,
    content: str,
    tools: list[dict],
    tool_executor: Callable[[str, dict], Awaitable[Any]],
    model: str = "claude-sonnet-4-5-20251101",
    max_tokens: int = 2000,
    system_prompt: str = "",
) -> AsyncIterator[str | ToolCallEvent]:
    _require_key()

    thread = _threads.get(thread_id)
    if not thread:
        thread = {"system_prompt": system_prompt, "messages": []}
        _threads[thread_id] = thread
    elif system_prompt and not thread.get("system_prompt"):
        thread["system_prompt"] = system_prompt

    messages: list[dict] = list(thread["messages"]) + [{"role": "user", "content": content}]
    client = anthropic.AsyncAnthropic(api_key=CLAUDE_KEY)

    while True:
        resp = await client.messages.create(
            model=model,
            max_tokens=max_tokens,
            system=thread["system_prompt"] or None,
            messages=messages,
            tools=tools,  # type: ignore[arg-type]
        )

        text_parts: list[str] = []
        tool_use_blocks: list[Any] = []
        for block in resp.content:
            if block.type == "text":
                text_parts.append(block.text)
            elif block.type == "tool_use":
                tool_use_blocks.append(block)

        if resp.stop_reason == "tool_use":
            for text in text_parts:
                chunk_size = 30
                for i in range(0, len(text), chunk_size):
                    yield text[i : i + chunk_size]

            content_dicts: list[dict] = []
            for block in resp.content:
                if block.type == "text":
                    content_dicts.append({"type": "text", "text": block.text})
                elif block.type == "tool_use":
                    content_dicts.append({
                        "type": "tool_use",
                        "id": block.id,
                        "name": block.name,
                        "input": dict(block.input),
                    })

            tool_results: list[dict] = []
            for block in tool_use_blocks:
                yield ToolCallEvent(
                    name=block.name,
                    input=dict(block.input),
                    call_id=block.id,
                    result=None,
                )
                result = await tool_executor(block.name, dict(block.input))
                yield ToolCallEvent(
                    name=block.name,
                    input=dict(block.input),
                    call_id=block.id,
                    result=result,
                )
                tool_results.append({
                    "type": "tool_result",
                    "tool_use_id": block.id,
                    "content": json.dumps(result, default=str),
                })

            messages.append({"role": "assistant", "content": content_dicts})
            messages.append({"role": "user", "content": tool_results})
        else:
            final_text = "".join(text_parts)
            chunk_size = 30
            for i in range(0, len(final_text), chunk_size):
                yield final_text[i : i + chunk_size]

            thread["messages"].append({"role": "user", "content": content})
            thread["messages"].append({"role": "assistant", "content": final_text})
            break
